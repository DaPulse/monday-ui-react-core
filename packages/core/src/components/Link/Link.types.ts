export type LinkTarget = "_blank" | "_self" | "_parent" | "_top";

export type IconPosition = "start" | "end";
