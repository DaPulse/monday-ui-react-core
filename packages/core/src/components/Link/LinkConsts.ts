/**
 * @deprecated
 */
export enum LinkTarget {
  NEW_WINDOW = "_blank",
  SELF = "_self",
  PARENT = "_parent",
  TOP = "_top"
}

/**
 * @deprecated
 */
export enum IconPosition {
  START = "start",
  END = "end"
}
