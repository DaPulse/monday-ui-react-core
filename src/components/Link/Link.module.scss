@import "../../styles/themes";
@import "../../styles/typography";
@import "../../styles/global-css-settings";
@import "../../styles/states";

.mondayStyleLink {
  display: flex;
  align-items: center;
  @include theme-prop(color, link-color);
  @include font-link();
  text-decoration: none;
  writing-mode: horizontal-tb;
  @include focus-style();
}

.mondayStyleLink:hover .mondayStyleLinkText {
  text-decoration: underline;
}

.mondayStyleLink.inlineText {
  display: inline-flex;
  align-items: unset;
}

.mondayStyleLink.inheritFontSize {
  font-size: inherit;
  line-height: inherit;
}

.mondayStyleLinkIconStart {
  line-height: 24px;
  margin-right: $spacing-small;
}

.mondayStyleLinkIconEnd {
  line-height: 24px;
  margin-left: $spacing-small;
}

@supports (margin-inline-start: initial) {
  .mondayStyleLinkIconStart {
    line-height: 24px;
    margin-right: 0;
    margin-inline-end: $spacing-small;
  }
  .mondayStyleLinkIconEnd {
    line-height: 24px;
    margin-left: 0;
    margin-inline-start: $spacing-small;
  }
}
