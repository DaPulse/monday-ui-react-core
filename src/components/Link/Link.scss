@import "../../styles/themes.scss";
@import "../../styles/typography.scss";
@import "../../styles/global-css-settings.scss";
@import "../../styles/states.scss";

.monday-style-link {
  display: flex;
  align-items: center;
  @include theme-prop(color, link-color);
  @include font-link();
  text-decoration: none;
  writing-mode: horizontal-tb;
  @include focus-style();
  &:hover {
    .monday-style-link--text {
      text-decoration: underline;
    }
  }
  &--icon-start {
    line-height: 24px;
    margin-right: $spacing-small;
  }
  &--icon-end {
    line-height: 24px;
    margin-left: $spacing-small;
  }
}

@supports (margin-inline-start: initial) {
  .monday-style-link {
    &--icon-start {
      line-height: 24px;
      margin-right: 0;
      margin-inline-end: $spacing-small;
    }
    &--icon-end {
      line-height: 24px;
      margin-left: 0;
      margin-inline-start: $spacing-small;
    }
  }
}
