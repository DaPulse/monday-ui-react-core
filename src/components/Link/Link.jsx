import React, { forwardRef, useCallback } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import NOOP from "lodash/noop";
import "./Link.scss";
import { LINK_TARGET, ICON_POSITION } from "./LinkConsts";
import Icon from "../Icon/Icon";

const Link = forwardRef(
  (
    {
      componentClassName,
      href,
      text,
      rel,
      onClick,
      target,
      ariaLabelDescription,
      icon,
      iconPosition,
      id,
      ariaLabeledBy,
      disableNavigation
    },
    ref
  ) => {
    const isStart = iconPosition === ICON_POSITION.START;

    const onClickWrapper = useCallback(
      e => {
        if (disableNavigation) {
          e.preventDefault();
        }
        onClick && onClick(e);
      },
      [disableNavigation, onClick]
    );

    return (
      <a
        id={id}
        href={href}
        rel={rel}
        ref={ref}
        onClick={onClickWrapper}
        target={target}
        className={cx("monday-style-link", componentClassName)}
        aria-label={ariaLabelDescription}
        aria-labelledby={ariaLabeledBy}
      >
        {getIcon(isStart, icon, "monday-style-link--icon-start")}
        <span className="monday-style-link--text">{text}</span>
        {getIcon(!isStart, icon, "monday-style-link--icon-end")}
      </a>
    );
  }
);

function getIcon(shouldShow, icon, className) {
  if (!shouldShow) return;
  return <Icon className={className} clickable={false} icon={icon} iconType={Icon.type.ICON_FONT} />;
}

Link.target = LINK_TARGET;
Link.position = ICON_POSITION;

Link.propTypes = {
  componentClassName: PropTypes.string,
  href: PropTypes.string,
  text: PropTypes.string,
  rel: PropTypes.string,
  onClick: PropTypes.func,
  target: PropTypes.oneOf([Link.target.NEW_WINDOW, Link.target.PARENT, Link.target.SELF, Link.target.TOP]),
  /** Aria label description */
  ariaLabelDescription: PropTypes.string,
  /** element id to describe the counter accordingly */
  ariaLabeledBy: PropTypes.string,
  icon: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  iconPosition: PropTypes.oneOf([Link.position.START, Link.position.END]),
  id: PropTypes.string,
  disableNavigation: PropTypes.bool
};

Link.defaultProps = {
  /**
   * Id to add to the
   */
  id: "",
  /**
   * Classname to be added to the component container
   */
  componentClassName: "",
  /**
   * the href to link the component to
   */
  href: "",
  /**
   * the link text
   */
  text: "",
  /**
   * The rel attribute defines the relationship between a linked resource and the current document
   */
  rel: "noreferrer",
  /**
   * on link click callback
   */
  onClick: NOOP,
  /**
   * the target window where the link should be open
   */
  target: Link.target.NEW_WINDOW,
  /**
   * Aria label
   */
  ariaLabelDescription: "",
  /**
   * icon to add to the link
   */
  icon: "",
  /**
   * where the icon should be located
   */
  iconPosition: Link.position.START,
  /**
   * Id to add to the link element
   */
  ariaLabeledBy: "",
  /**
   * Disable navigation on click
   */
  disableNavigation: false
};

export default Link;
