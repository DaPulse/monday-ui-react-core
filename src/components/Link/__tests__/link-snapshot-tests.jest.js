import React from "react";
import renderer from "react-test-renderer";
import Link from "../Link";

describe("Link renders correctly", () => {
  it("with empty props", () => {
    const tree = renderer.create(<Link />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with id", () => {
    const tree = renderer.create(<Link id="id" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with className", () => {
    const tree = renderer.create(<Link componentClassName="my-classname" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with href", () => {
    const tree = renderer.create(<Link href="http://somelink.com" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with rel", () => {
    const tree = renderer.create(<Link rel="something" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with onClick", () => {
    const tree = renderer.create(<Link onClick={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with target", () => {
    const tree = renderer.create(<Link target={Link.target.TOP} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("a11y", () => {
    const tree = renderer.create(<Link ariaLabelDescription="My Link" ariaLabeledBy="anotherId" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
