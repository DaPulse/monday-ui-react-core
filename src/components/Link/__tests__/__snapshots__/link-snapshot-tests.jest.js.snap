// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly a11y 1`] = `
<a
  aria-label="My Link"
  aria-labelledby="anotherId"
  className="monday-style-link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with className 1`] = `
<a
  aria-label=""
  aria-labelledby=""
  className="monday-style-link my-classname"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with empty props 1`] = `
<a
  aria-label=""
  aria-labelledby=""
  className="monday-style-link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with href 1`] = `
<a
  aria-label=""
  aria-labelledby=""
  className="monday-style-link"
  href="http://somelink.com"
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with id 1`] = `
<a
  aria-label=""
  aria-labelledby=""
  className="monday-style-link"
  href=""
  id="id"
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with onClick 1`] = `
<a
  aria-label=""
  aria-labelledby=""
  className="monday-style-link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with rel 1`] = `
<a
  aria-label=""
  aria-labelledby=""
  className="monday-style-link"
  href=""
  id=""
  onClick={[Function]}
  rel="something"
  target="_blank"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;

exports[`Link renders correctly with target 1`] = `
<a
  aria-label=""
  aria-labelledby=""
  className="monday-style-link"
  href=""
  id=""
  onClick={[Function]}
  rel="noreferrer"
  target="_top"
>
  <span
    className="monday-style-link--text"
  >
    
  </span>
</a>
`;
